lm_eval/tasks/custom_metrics/multiple_metrics/evaluation.py

```diff
@@ -0,0 +1,107 @@
+import json
+import os
+import time
+from concurrent.futures import ThreadPoolExecutor
+from pathlib import Path
+from threading import Lock
+from typing import Optional
+
+from .containerized_eval import eval_string_script
+
+# Get working directory
+WORKING_DIR = Path(__file__).parent.parent
+
+# program: str => Result
+CACHE = dict()
+CACHE_LOCK = Lock()
+
+
+def cache_get(program: str) -> Optional[dict]:
+    if program in CACHE:
+        result = CACHE[program]
+        return result
+    else:
+        return None
+
+
+def cache_set(program: str, result: dict):
+    if program in CACHE:
+        print("Setting already-existing cache")
+    CACHE[program] = result
+
+
+def cached_eval_script(problem, index) -> dict:
+    # here prompt is already included in completions
+    program = problem["completions"][index] + "\n" + problem["tests"]
+    CACHE_LOCK.acquire(True)
+    cached = cache_get(program)
+    if cached is not None:
+        CACHE_LOCK.release()
+        return cached
+    else:
+        result_yaml = dict()
+        cache_set(program, result_yaml)
+        CACHE_LOCK.release()
+        result_dict = eval_string_script(problem["language"], program)
+        for k in result_dict.keys():
+            result_yaml[k] = result_dict[k]
+            result_yaml["timestamp"] = int(time.time())
+        return result_yaml
+
+
+def get_test_results_json_path(
+    output_dir: str, problem_json_path: str, input_dir: Path
+) -> Path:
+    suffixes = ".results.json"
+    problem_name = problem_json_path[: -len(".json")]
+    if input_dir:
+        raise ValueError("input dir given")
+        return Path(output_dir) / (
+            problem_json_path.relative_to(input_dir).parent / (problem_name + suffixes)
+        )
+    return Path(output_dir) / (problem_name + suffixes)
+
+
+def evaluate_problem(
+    output_dir: str, problem_json_path: str, max_workers: int, input_dir: Path = None
+):
+    with open(problem_json_path, "r") as f:
+        problem = json.load(f)
+    if len(problem["completions"]) == 0:
+        return
+    test_results_path = get_test_results_json_path(
+        output_dir, problem_json_path, input_dir
+    )
+    test_results_path.parent.mkdir(mode=0o755, parents=True, exist_ok=True)
+
+    # TODO: remove or fix, code is commented because:
+    # if results exist from any previous run they will be taken mistakingly in the else?
+    # if not test_results_path.exists():
+    test_results = problem.copy()
+    del test_results["completions"]
+    test_results["results"] = []
+    # else:
+    #    with open(test_results_path, "r") as f:
+    #        test_results = json.load(f)
+
```

Comment:
Regarding the TODO above, I think what's happening is OK since `output_dir` is a new temporary directory.
When we were building MultiPL-E, we would reuse the same output directory for a particular experiment configuration, and this code helped reuse results.

The only problem with using a temporary directory is that one may want to debug the generated problem JSON file. But, that can be done in the original MultiPL-E code.

Reply:
This setup actually also saves the generations in a file `generations.json`, in a list with the same order as humaneval problems which can be used for debugging or running evaluation again, the `temp_dir` is just duplication to use the same code as MultiPL-E. 

Regarding saving `test_results` do you usually use it when execution fails and evaluates only part of the problems, so you reload those results and continue evaluation for the rest? If that's the case how often does this happen? If it's common we can add a workaround to keep the feature.
Otherwise if evaluation is done for all problems saving `test_results` will be as good as saving the scores of the experiment if I'm not mistaken, since `pass@k` computation is fast.